### PR TITLE
rustdoc: Fix ICE for intra-doc link on intermediate re-export

### LINKED ISSUE #109282: Fix ICE for intra-doc link on intermediate re-export
```rust
mod m {
    pub enum ValueEnum {}
}
mod m2 {
    /// [`ValueEnum`]
    pub	use crate::m::ValueEnum;
}
pub use m2::ValueEnum;
```

It needs to generalize the fix implemented in https://github.com/rust-lang/rust/pull/109266.

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes https://github.com/rust-lang/rust/issues/109282.

This PR is based on #109266 as it includes its commit to make this work.

@petrochenkov: It was exactly as you predicted, adding the `DefId` to the attributes fixed the error for intermediate re-exports as well. Thanks a lot!

r? @petrochenkov

### CODE REVIEW

**Comment by petrochenkov** [comment]
Blocked on https://github.com/rust-lang/rust/pull/109312.

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/109312 has landed.
@rustbot author

**Comment by GuillaumeGomez** [comment]
Rebased.

@bors ready

File: src/librustdoc/clean/types.rs [hunk 1]
```diff
@@ -980,6 +969,38 @@ impl AttributesExt for [ast::Attribute] {
...
+            .filter_map(ast::Attribute::meta_item_list)
+            .flatten()
+    }
+
+    fn iter<'a>(&'a self) -> Self::Attributes<'a> {
+        self.into_iter()
+    }
+}
+
+impl AttributesExt for [(ast::Attribute, Option<DefId>)] {
```
Comment (petrochenkov) [inline src/librustdoc/clean/types.rs:988]: This impl forces attribute cloning everywhere.
See how `Attributes::from_ast_iter` is implemented - it uses `(&'a ast::Attribute, Option<DefId>)` to use attribute references pointing to tcx arenas without cloning.

(Also, I'm not sure this `AttributesExt` trait is useful in general, it's never used generically, only for providing method syntax.)

Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:988]: Good points. I'll see how to improve all that.

**Comment by GuillaumeGomez** [comment]
I used `Cow` to avoid clones as much as possible. It's still necessary in one place (where we edit an attribute) but it's the only place. It also allowed to remove the need to create temporary `Vec`, which is quite nice.

As for `AttributesExt`: its `cfg` method is used in a lot of different places, which in itself isn't a problem, except that if the function was moved out of the trait, it would require to instead either:
 * Write two functions, each taking a different kind of iterable data.
 * Take two callbacks to get the iterators.

I think none of these options is better than the current status so I preferred to keep it as is. One think I did though was to remove its `span` associated function since it was used only in place, so no need to keep it there.

**Comment by petrochenkov** [comment]
@GuillaumeGomez 
Could you check whether this PR fixes the linked issues - https://github.com/rust-lang/rust/issues/109631 https://github.com/rust-lang/rust/issues/109614 https://github.com/rust-lang/rust/issues/109424?
(And add "fixes" annotations to the PR description if they do.)

File: src/librustdoc/clean/mod.rs [hunk 2]
```diff
@@ -2168,32 +2168,39 @@ impl<'hir> hir::intravisit::Visitor<'hir> for OneLevelVisitor<'hir> {
...
+    let mut attributes: Vec<(Cow<'hir, ast::Attribute>, Option<DefId>)> = Vec::new();
     let mut first = true;
-    let hir_map = tcx.hir();
+    let hir_map = cx.tcx.hir();
     let mut visitor = OneLevelVisitor::new(hir_map, target_def_id);
     let mut visited = FxHashSet::default();
 
     // If the item is an import and has at least a path with two parts, we go into it.
     while let hir::ItemKind::Use(path, _) = item.kind && visited.insert(item.hir_id()) {
+        let import_parent = cx.tcx.opt_local_parent(prev_import).map(|def_id| def_id.to_def_id());
```
Comment (petrochenkov) [inline src/librustdoc/clean/mod.rs:2184]: @GuillaumeGomez 
This PR wasn't rebased correctly, after https://github.com/rust-lang/rust/pull/109312 doc fragment contain ids of imports themselves, not their parent modules.
This is most likely the reason why this PR doesn't fix the `rustix` issues it was supposed to fix.

Comment (GuillaumeGomez) [reply src/librustdoc/clean/mod.rs:2184]: Oh. It's surprising that the test I added was still working then. Well, great that your PR fixed it.

**Pull request merged**